### DISCUSSION TITLE
Voilà with JupyterLite

### DISCUSSION THREAD

**Discussion by jtpio** [jupyterlite/jupyterlite#202] [Ideas]
At some point in the future, it would be interesting to look into having an in-browser Voilà as well.

Just like with today's JupyterLab + Voilà, folks could then work on their notebooks in JupyterLite, and then hit the `/voila` endpoint to render their application as a dashboard.

One advantage of this would be the scaling and hosting of dashboards as it would also be a bunch of static files just like JupyterLite.

Not sure there would be much to reuse from the existing Voila Python backend. Probably we'll at least need Voila to be a lab-based application first (https://github.com/voila-dashboards/voila/pull/846). And then plug the `ServiceManager` and `serverlite` machinery. And the [kernels service](https://github.com/jtpio/jupyterlite/blob/main/packages/kernel/src/kernels.ts) could be the one taking of not accepting `execute_requests`.

Also the whole nbconvert templating would probably need a different approach.

**Comment by bollwyvl** [comment]
> Also the whole nbconvert templating would probably need a different approach.

Right: much like with #141, it's insane to fire up 20mb of python, a webworker to do something that the browser can be coaxed into doing with `nunjucks`

> need Voila to be a lab-based application

Or do we? The voila frontend could just _be_ an on-demand lite site, as #45 is table stakes _anyway_ for reproducible, archival-grade self-hosting. This was part of the motivation of the (still mostly untested) `jupyter-lite.ipynb` format (which could be extended to any-old-notebook) as it would actually allow one to capture required labextensions in a semi-automated way in the (not yet started) #165. In a more controlled deployment environment, all of the wheelhouses, federated_extensions, etc. could be shared files on disk, and extensions wouldn't have to be actually installed, could be sourced from internal `pypi` mirrors, etc. Perhaps a reusable `LiteHandler` handler needs to be part of #41, with the `config` replacing references to `manager` all over so it can support multiple builds.

In the `voilite` serverlite extension, the browser would have access to all of the underlying JSON, but the real one could continue to do the server-side rendering for keeping stuff private.

**Comment by seidlr** [comment]
It would be really great to be able to build some interactive widgets with bqplot, ipyvuetify etc. in a notebook and then only serve the  widget on a different website using voila-embed without the need of running a server.

I currently go through the trouble of deploying voila dashboards as app on my own CapRover server (similar to Heroku) and then adding the interactive widgets using voila-embed. E.g. [here](https://blog.apps.talksportsdata.com/tennis_murray.html)

**Reply by deeplook** [reply]
Just curious about any updates on this...?

**Comment by bollwyvl** [comment]
There's this more recent issue:  https://github.com/jupyterlite/jupyterlite/issues/608
